server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -188,11 +215,12 @@ public DataStream(
         this.indexMode = indexMode;
         this.lifecycle = lifecycle;
         this.failureStoreEnabled = failureStoreEnabled;
+        assert backingIndices.indices.isEmpty() == false;
+        assert assertConsistent(backingIndices.indices);
+        assert replicated == false || backingIndices.rolloverOnWrite == false
```

Comment:
Should we check here that the failure indices aren't also marked for rollover on write? I'm in the middle on this: It's less of a problem for failure store indices since we synthesize a template for them and so the risk of not being able to rollover the indices after promoting them is low. That said, I don't think we should allow it on one set of indices and not the other.